File: python/ENEPro/EneQuotaUtil.py
# -*- coding: UTF-8 -*- 
'''
Created on 2016年12月21日

@author: Administrator
'''
from EneParameter import EneParameter
from StockEneInfo import StockEneInfo
from CommonUtil import CommonUtil
from MySQLUtil import MySQLUtil
from EneParameterRange import EneParameterRange
class EneQuotaUtil :
    
    global INFOLIST
    INFOLIST=[]
    '''
     * 获取一个股票在指定ENE参数内的适宜程度
     * 
     * @param eneParameter
     *            ENE参数
     * @return 适宜程度
    '''
    def get_Standard_Frequency(self,eneParam):
        times = 0
        mysqlutil = MySQLUtil()
        INFOLIST.extend(mysqlutil.getStockEneInfos(eneParam.getCode())) 
        length = len(INFOLIST);
#         print(length)
        
        for i in range(length):
            index = length -1 - i
            info = INFOLIST[index]
            avgClose = self.getAvgClose(info,index,eneParam.getDays())
            
            upper = self.getUpperIndex(eneParam.getUpper(),avgClose)
            lower = self.getLowerIndex(eneParam.getLower(),avgClose)
#             print(upper)
            if info.getHigh() <= upper and info.getLow() >= lower:
                times=times+1
       
        frequency = float(times)/length
        # lengthprint frequency
        return frequency
    
    
    '''
     * 获取一个股票最适宜的ENE参数
     * 
     * @param stockCode
     *            股票代号
     * @return 最适宜的ENE参数
    '''
    def get_Optimum_Param(self,r):
        frequencyList = []
        paramList = []
        for i in range(r.getUpperMin(),r.getUpperMax()+1):
            for j in range(r.getLowerMin(),r.getLowerMax()+1):
                for k in range(r.getDaysMin(),r.getDaysMax()+1):
                    # print str(i)+"---"+str(j)+"---"+str(k)
                    param = EneParameter()
                    param.setCode(r.getCode())
                    param.setDays(k)
                    param.setLower(j)
                    param.setUpper(i)
                    frequency = self.get_Standard_Frequency(param)
                    # print frequency
                    frequencyList.append(frequency)
                    paramList.append(param)
            
        commonutil = CommonUtil()
        indexList = commonutil.getArrayListMax(frequencyList)
        optimumParamList = []
        for index in indexList:
            param = paramList[index]
            optimumParamList.append(param)
            print(frequencyList[index])
            print(param.getCode())
            print(param.getUpper())
            print(param.getLower())
            print(param.getDays())
            print '---------------------------'
        return optimumParamList
        
    '''
     * 获取ENE指标上限
     * 
     * @param upper
     *            偏移程度
     * @param avgClose
     *            平均收盘价
     * @return
    '''
    def getUpperIndex(self,upper,avgClose) :
        floatDegree = float(upper)/100
        upperIndex = (1 + floatDegree) * avgClose
        return upperIndex

    '''
     * 获取ENE指标下限
     * 
     * @param lower
     *            偏移程度
     * @param avgClose
     *            平均收盘价
     * @return
    '''
    def getLowerIndex(self,lower,avgClose) :
        floatDegree = float(lower)/100
        lowerIndex = (1 - floatDegree) * avgClose
        return lowerIndex
    
    '''
     * 获取n天股票收盘价的均值
     * 
     * @param stockEneInfo
     *            股票数据
     * @param index
     *            股票数据在股票数据集合中的下标
     * @param days
     *            n天
     * @return 收盘价均值
    '''
    def getAvgClose(self,stockEneInfo, index, days) :
        date = stockEneInfo.getDate()
        commonUtil=CommonUtil()
        dateBefore=commonUtil.getDaysAfter(date, -days)
        closeList=[]
        allClose=0.0
#         print(len(INFOLIST))
        for i in range(0,days) :
            infoBefore=StockEneInfo()
            if index >= i :
                infoBefore = INFOLIST[index - i]    
            else :
                break
            
            if infoBefore.getDate()>= dateBefore :
                closeList.append(infoBefore.getClose());
        
        for close in closeList:
            allClose += close
        
        avgClose = allClose / len(closeList)
        return avgClose    


# p = EneQuotaUtil()
# eneparam = EneParameterRange()
# eneparam.setCode("sh600011")
# eneparam.setDaysMax(11)
# eneparam.setDaysMin(9)
# eneparam.setLowerMax(10)
# eneparam.setLowerMin(8)
# eneparam.setUpperMax(12)
# eneparam.setUpperMin(10)
# p.get_Optimum_Param(eneparam)

File: java/eneProject/src/org/nju/ene/util/EneQuotaUtil.java
package org.nju.ene.util;

import java.util.Date;
import java.util.ArrayList;

import org.nju.ene.pojo.EneParameter;
import org.nju.ene.pojo.EneParameterRange;
import org.nju.ene.pojo.StockEneInfo;

public class EneQuotaUtil {
	private static ArrayList<StockEneInfo> infoList;
	private static EneParameter parameter;

	/**
	 * 获取一个股票在指定ENE参数内的适宜程度
	 * 
	 * @param eneParameter
	 *            ENE参数
	 * @return 适宜程度
	 */
	public static double getStandardFrequency(EneParameter eneParameter) {
		int times = 0;
		double frequency = 0;
		infoList = MySQLUtil.getStockEneInfos(eneParameter.getCode());
		int length = infoList.size();
	//	System.out.println(length);
		
		for(int i=0;i<length;i++){
			int index = length - 1 - i;
			StockEneInfo info = infoList.get(index);
			double avgClose = getAvgClose(info, index, eneParameter.getDays());
			double upper = getUpperIndex(eneParameter.getUpper(), avgClose);
			double lower = getLowerIndex(eneParameter.getLower(), avgClose);
		//	System.out.println(avgClose+": "+upper+": "+info.getHigh()+": "+lower+": "+info.getLow());
			if(info.getHigh()<=upper && info.getLow()>=lower)
				times++;
		}
		//System.out.println(times);
		frequency = (double)times/length;
		return frequency;
	}

	/**
	 * 获取一个股票最适宜的ENE参数
	 * 
	 * @param stockCode
	 *            股票代号
	 * @return 最适宜的ENE参数
	 */
	public static ArrayList<EneParameter> getOptimumParameter(EneParameterRange range) {
		ArrayList<Double> frequencyList = new ArrayList<>();
		ArrayList<EneParameter> parameterList = new ArrayList<>();
		for(int i=range.getUpperMin();i<=range.getUpperMax();i++){
			for(int j=range.getLowerMin();j<=range.getLowerMax();j++){
				for(int k=range.getDaysMin();k<=range.getDaysMax();k++){
					EneParameter parameter = new EneParameter();
					parameter.setCode(range.getCode());
					parameter.setDays(k);
					parameter.setLower(j);
					parameter.setUpper(i);
					double frequecy = getStandardFrequency(parameter);
					frequencyList.add(frequecy);
					parameterList.add(parameter);
					System.out.println(k+"---"+j+"---"+i);
				}
			}
		}
		ArrayList<Integer> indexList = CommonUtil.getArrayListMax(frequencyList);
		ArrayList<EneParameter> optimumParameterList = new ArrayList<>();
		for(int index: indexList){
			EneParameter parameter = parameterList.get(index);
			optimumParameterList.add(parameterList.get(index));
			System.out.println(frequencyList.get(index));
			System.out.println(parameter.getCode()+"------ upper: "+parameter.getUpper()+"---- lower: "+parameter.getLower()+"---- days: "+parameter.getDays());
		}
		return optimumParameterList;
	}

	/**
	 * 获取ENE指标上限
	 * 
	 * @param upper
	 *            偏移程度
	 * @param avgClose
	 *            平均收盘价
	 * @return
	 */
	private static double getUpperIndex(int upper, double avgClose) {
		double floatDegree = (double)upper / 100;
	//	System.out.println("float: "+floatDegree);
		double upperIndex = (1 + floatDegree) * avgClose;
	//	System.out.println("avg: "+avgClose+"Upper: "+upperIndex);
		return upperIndex;
	}

	/**
	 * 获取ENE指标下限
	 * 
	 * @param lower
	 *            偏移程度
	 * @param avgClose
	 *            平均收盘价
	 * @return
	 */
	private static double getLowerIndex(int lower, double avgClose) {
		double floatDegree = (double)lower / 100;
		double lowerIndex = (1 - floatDegree) * avgClose;
		return lowerIndex;
	}

	/**
	 * 获取n天股票收盘价的均值
	 * 
	 * @param stockEneInfo
	 *            股票数据
	 * @param index
	 *            股票数据在股票数据集合中的下标
	 * @param days
	 *            n天
	 * @return 收盘价均值
	 */
	private static double getAvgClose(StockEneInfo stockEneInfo, int index, int days) {
		Date date = stockEneInfo.getDate();
//		System.out.println(index+"----"+date);
		Date dateBefore = CommonUtil.getDaysAfter(date, -days);
		ArrayList<Double> closeList = new ArrayList<>();
		double allClose = 0;

		for (int i = 0; i < days; i++) {
			StockEneInfo infoBefore = null;
			if (index >= i) {
				infoBefore = infoList.get(index - i);	
			}else{
				break;
			}
		//	System.out.print(date+"------"+infoBefore.getDate() +"------"+dateBefore+"/////");
			if (infoBefore.getDate().getTime() >= dateBefore.getTime()) {
		//		System.out.println(date+"------"+infoBefore.getDate() +"------"+dateBefore);
		//		System.out.println(infoBefore.getClose());
				closeList.add(infoBefore.getClose());
			}
		}

		for (double close : closeList) {
			allClose += close;
		}
		double avgClose = allClose / closeList.size();
	//	System.out.println(allClose+"---"+avgClose);
		return avgClose;
	}
}
